module-system/module-implementations/sov-sequencer-registry/src/lib.rs

```diff
@@ -26,31 +26,86 @@ pub struct Sequencer<C: sov_modules_api::Context> {
     #[module]
     pub(crate) bank: sov_bank::Bank<C>,
 
-    /// The sequencer address on the rollup.
-    #[state]
-    pub(crate) seq_rollup_address: StateValue<C::Address>,
-
-    /// The sequencer address on the DA.
+    /// Only batches from sequencers from this list are going to be processed
+    ///
     #[state]
-    pub(crate) seq_da_address: StateValue<Vec<u8>>,
+    pub(crate) allowed_sequencers: StateMap<Vec<u8>, C::Address>,
 
     /// Coin's that will be slashed if the sequencer is malicious.
     /// The coins will be transferred from `self.seq_rollup_address` to `self.address`
-    /// and locked forever.
+    /// and locked forever, until sequencer decides to exit
+    /// Only sequencers in `allowed_sequencers` list are allowed to exit.
     #[state]
     pub(crate) coins_to_lock: StateValue<sov_bank::Coins<C>>,
 }
 
-impl<C: sov_modules_api::Context> sov_modules_api::Module for Sequencer<C> {
+/// Result of applying blob, from sequencer point of view.
+pub enum SequencerOutcome {
+    Completed,
+    Slashed { sequencer: Vec<u8> },
+}
+
+impl<C: sov_modules_api::Context> sov_modules_api::Module for SequencerRegistry<C> {
     type Context = C;
 
     type Config = SequencerConfig<C>;
 
+    type CallMessage = call::CallMessage;
+
     fn genesis(
         &self,
         config: &Self::Config,
         working_set: &mut WorkingSet<C::Storage>,
     ) -> Result<(), Error> {
         Ok(self.init_module(config, working_set)?)
     }
+
+    fn call(
+        &self,
+        message: Self::CallMessage,
+        context: &Self::Context,
+        working_set: &mut WorkingSet<<Self::Context as Spec>::Storage>,
+    ) -> Result<CallResponse, Error> {
+        Ok(match message {
+            call::CallMessage::Register { da_address } => {
+                self.register(da_address, context, working_set)?
+            }
+            call::CallMessage::Exit { da_address } => {
+                self.exit(da_address, context, working_set)?
+            }
+        })
+    }
+}
+
+impl<C: sov_modules_api::Context> SequencerRegistry<C> {
```

Comment:
Let's move the code to `call.rs`?

Reply:
Could you please elaborate why? 

Then it makes genesis depending on "call.rs". And `get_coins_to_lock` does not have corresponding call nor query message, it is just helper method.

Reply:
My mistake I meant only `register_sequencer` which is used only in `call.rs`.

Reply:
`register_sequencer` is also used in `genesis`. Originally whole logic was just in `call.rs`, but as I adjusted genesis, I've moved it to lib.rs